File: src/data/spreads.js
// Celtic Cross traditional layout positions (relative grid coordinates)
const CELTIC_LAYOUT = [
  { x: 0, y: 0, rotate: 0 },      // 1: Present (center)
  { x: 0, y: 0, rotate: 90 },     // 2: Challenge (crossing)
  { x: -1, y: 0, rotate: 0 },     // 3: Past (left)
  { x: 1, y: 0, rotate: 0 },      // 4: Future (right)
  { x: 0, y: -1, rotate: 0 },     // 5: Above (crown)
  { x: 0, y: 1, rotate: 0 },      // 6: Below (foundation)
  { x: 2.2, y: 1.5, rotate: 0 },  // 7: Advice (staff bottom)
  { x: 2.2, y: 0.5, rotate: 0 },  // 8: External (staff)
  { x: 2.2, y: -0.5, rotate: 0 }, // 9: Hopes/Fears (staff)
  { x: 2.2, y: -1.5, rotate: 0 }, // 10: Outcome (staff top)
]

export const SPREADS = {
  single: {
    name: "Daily Draw",
    positions: [{ name: "Today's Guidance", description: "Your guidance for today" }]
  },
  threeCard: {
    name: "Past · Present · Future",
    positions: [
      { name: "Past", description: "What has led to this moment" },
      { name: "Present", description: "Where you are now" },
      { name: "Future", description: "Where this path leads" }
    ]
  },
  situation: {
    name: "Situation · Action · Outcome",
    positions: [
      { name: "Situation", description: "The nature of what you face" },
      { name: "Action", description: "The approach advised" },
      { name: "Outcome", description: "The likely result" }
    ]
  },
  celtic: {
    name: "Celtic Cross",
    layout: CELTIC_LAYOUT,
    positions: [
      { name: "Present", description: "Current situation" },
      { name: "Challenge", description: "Obstacle you face" },
      { name: "Past", description: "Recent influences" },
      { name: "Future", description: "What's coming" },
      { name: "Above", description: "Best outcome" },
      { name: "Below", description: "Subconscious" },
      { name: "Advice", description: "How to approach" },
      { name: "External", description: "Outside influences" },
      { name: "Hopes/Fears", description: "Inner conflict" },
      { name: "Outcome", description: "Final result" }
    ]
  },
  horseshoe: {
    name: "Horseshoe",
    positions: [
      { name: "Past", description: "What has shaped the situation" },
      { name: "Present", description: "Current circumstances" },
      { name: "Hidden Influences", description: "Factors not immediately apparent" },
      { name: "Obstacles", description: "Challenges to overcome" },
      { name: "External Influences", description: "People or events affecting you" },
      { name: "Advice", description: "Guidance for moving forward" },
      { name: "Outcome", description: "Where this path leads" }
    ]
  }
}

File: training/scripts/spreads.py
"""
Tarot spread definitions for training data generation.

Mirrors the spread data from src/data/spreads.js.
"""

from dataclasses import dataclass
from typing import List, Dict
import random

@dataclass
class Position:
    name: str
    description: str


@dataclass
class Spread:
    id: str
    name: str
    positions: List[Position]
    weight: float  # Distribution weight for sampling


SPREADS: Dict[str, Spread] = {
    "single": Spread(
        id="single",
        name="Daily Draw",
        positions=[
            Position("Today's Guidance", "Your guidance for today")
        ],
        weight=0.15,  # Single card readings are common but simple
    ),

    "threeCard": Spread(
        id="threeCard",
        name="Past · Present · Future",
        positions=[
            Position("Past", "What has led to this moment"),
            Position("Present", "Where you are now"),
            Position("Future", "Where this path leads"),
        ],
        weight=0.30,  # Most popular spread
    ),

    "situation": Spread(
        id="situation",
        name="Situation · Action · Outcome",
        positions=[
            Position("Situation", "The nature of what you face"),
            Position("Action", "The approach advised"),
            Position("Outcome", "The likely result"),
        ],
        weight=0.20,
    ),

    "horseshoe": Spread(
        id="horseshoe",
        name="Horseshoe",
        positions=[
            Position("Past", "What has shaped the situation"),
            Position("Present", "Current circumstances"),
            Position("Hidden Influences", "Factors not immediately apparent"),
            Position("Obstacles", "Challenges to overcome"),
            Position("External Influences", "People or events affecting you"),
            Position("Advice", "Guidance for moving forward"),
            Position("Outcome", "Where this path leads"),
        ],
        weight=0.15,
    ),

    "celtic": Spread(
        id="celtic",
        name="Celtic Cross",
        positions=[
            Position("Present", "Current situation"),
            Position("Challenge", "Obstacle you face"),
            Position("Past", "Recent influences"),
            Position("Future", "What's coming"),
            Position("Above", "Best outcome"),
            Position("Below", "Subconscious"),
            Position("Advice", "How to approach"),
            Position("External", "Outside influences"),
            Position("Hopes/Fears", "Inner conflict"),
            Position("Outcome", "Final result"),
        ],
        weight=0.20,  # Complex but classic
    ),
}


def get_spread(spread_id: str) -> Spread:
    """Get a spread by ID."""
    return SPREADS[spread_id]


def get_all_spreads() -> List[Spread]:
    """Return all spreads."""
    return list(SPREADS.values())


def sample_spread_weighted() -> Spread:
    """Sample a spread according to weights."""
    spreads = list(SPREADS.values())
    weights = [s.weight for s in spreads]
    return random.choices(spreads, weights=weights, k=1)[0]


def get_spread_stats() -> Dict:
    """Return statistics about spreads."""
    return {
        "total_spreads": len(SPREADS),
        "spreads": {
            s.id: {
                "name": s.name,
                "positions": len(s.positions),
                "weight": s.weight
            }
            for s in SPREADS.values()
        }
    }


if __name__ == "__main__":
    stats = get_spread_stats()
    print(f"Total spreads: {stats['total_spreads']}")
    print("\nSpreads:")
    for spread_id, info in stats['spreads'].items():
        print(f"  {info['name']}: {info['positions']} positions (weight: {info['weight']:.0%})")

    print("\n--- Sample weighted distribution (100 samples) ---")
    samples = [sample_spread_weighted().id for _ in range(100)]
    for spread_id in SPREADS:
        count = samples.count(spread_id)
        print(f"  {spread_id}: {count}%")
